src/modeling_library/switch/update.jl

```diff
@@ -0,0 +1,115 @@
+mutable struct SwitchUpdateState{T}
+    weight::Float64
+    score::Float64
+    noise::Float64
+    prev_trace::Trace
+    trace::Trace
+    index::Int
+    discard::ChoiceMap
+    updated_retdiff::Diff
+    SwitchUpdateState{T}(weight::Float64, score::Float64, noise::Float64, prev_trace::Trace) where T = new{T}(weight, score, noise, prev_trace)
+end
+
+function update_recurse_merge(prev_choices::ChoiceMap, choices::ChoiceMap)
+    prev_choice_submap_iterator = get_submaps_shallow(prev_choices)
+    prev_choice_value_iterator = get_values_shallow(prev_choices)
+    choice_submap_iterator = get_submaps_shallow(choices)
+    choice_value_iterator = get_values_shallow(choices)
+    new_choices = DynamicChoiceMap()
+    for (key, value) in prev_choice_value_iterator
+        key in keys(choice_value_iterator) && continue
+        set_value!(new_choices, key, value)
+    end
+    for (key, node1) in prev_choice_submap_iterator
+        if key in keys(choice_submap_iterator)
+            node2 = get_submap(choices, key)
+            node = update_recurse_merge(node1, node2)
+            set_submap!(new_choices, key, node)
+        else
+            set_submap!(new_choices, key, node1)
+        end
+    end
+    for (key, value) in choice_value_iterator
+        set_value!(new_choices, key, value)
+    end
+    sel, _ = zip(prev_choice_submap_iterator...)
+    comp = complement(select(sel...))
+    for (key, node) in get_submaps_shallow(get_selected(choices, comp))
+        set_submap!(new_choices, key, node)
+    end
+    return new_choices
+end
+
+function update_discard(prev_trace::Trace, choices::ChoiceMap, new_trace::Trace)
+    discard = choicemap()
+    prev_choices = get_choices(prev_trace)
+    for (k, v) in get_submaps_shallow(prev_choices)
+        isempty(get_submap(get_choices(new_trace), k)) && continue
+        isempty(get_submap(choices, k)) && continue
+        set_submap!(discard, k, v)
+    end
+    for (k, v) in get_values_shallow(prev_choices)
+        has_value(get_choices(new_trace), k) || continue
+        has_value(choices, k) || continue
+        set_value!(discard, k, v)
```

Comment:
I found this hard to read. I think that writing out the logic with explicit if-else would be clearer here.

```
if has_value(get_choices(new_trace), k) && has_value(choices, k)
    set_value!(discard, k, v)
end
```
But also, I think this should be changed to:
```
if (!has_value(get_choices(new_trace), k)) || has_value(choices, k)
    set_value!(discard, k, v)
end
```

Reply:
Yes - this is a lot simpler and easier to understand.

Reply:
There was actually an error here that you caught.

You should add a choice to the discard if:
1. The choice is in the old trace but not in the new trace.
2. The choice is in the old trace and (the choice is in the new trace and in the constraints).

Previously, I was checking for 2 but I wasn't covering 1.